File: impl/swift/Coia.swift
// COIA v2 reference implementation (Swift).
// Tables and vectors are generated; see ../../gencode.py.
import Foundation

enum Coia {
    static let TATWEEL: UInt32 = 0x0640
    static let ZWJ: UInt32 = 0x200D
    static let ZWNJ: UInt32 = 0x200C

    /// The reflexive sentinel, distinct from the empty string (§4.1).
    static let ME = "\u{0}COIA-ME"

    static func inRanges(_ cp: UInt32, _ rs: [(UInt32, UInt32)]) -> Bool {
        rs.contains { cp >= $0.0 && cp <= $0.1 }
    }

    // Default Case Folding from the normative table (Appendix B.6). Deliberately does
    // NOT use lowercased(): case mapping and case folding are different operations.
    static func casefold(_ s: String) -> String {
        var out = ""
        for u in s.unicodeScalars {
            if let f = FOLDGAP[u.value] { out += f } else { out.unicodeScalars.append(u) }
        }
        return out
    }

    static func joinerOk(_ cps: [Unicode.Scalar], _ i: Int) -> Bool {
        let cp = cps[i].value
        if cp == ZWJ { return i > 0 && inRanges(cps[i - 1].value, VIRAMA) }
        if cp == ZWNJ {
            if i > 0 && inRanges(cps[i - 1].value, VIRAMA) { return true }
            return i > 0 && i + 1 < cps.count
                && inRanges(cps[i - 1].value, ARABIC) && inRanges(cps[i + 1].value, ARABIC)
        }
        return false
    }

    static func isLetterOrNumber(_ gc: Unicode.GeneralCategory) -> Bool {
        switch gc {
        case .uppercaseLetter, .lowercaseLetter, .titlecaseLetter, .modifierLetter,
             .otherLetter, .decimalNumber, .letterNumber, .otherNumber: return true
        default: return false
        }
    }

    /// Reduce a string to COIA canonical form (§5).
    static func normalize(_ s: String) -> String {
        let nfkc = s.precomposedStringWithCompatibilityMapping
        let cps = Array(casefold(nfkc).unicodeScalars)
        var out = String.UnicodeScalarView()
        var onBase = false
        for (i, u) in cps.enumerated() {
            let cp = u.value
            let gc = u.properties.generalCategory
            if inRanges(cp, SPLIT) {
                out.append(" ")
                onBase = false
            } else if cp == TATWEEL {
                // decorative
            } else if gc == .format {
                if joinerOk(cps, i) { out.append(u) }
            } else if gc == .enclosingMark {
                // enclosing marks are decorative
            } else if gc == .nonspacingMark || gc == .spacingMark {
                if onBase { out.append(u) }
            } else if isLetterOrNumber(gc) {
                out.append(cp == 0x02BC ? Unicode.Scalar(0x02BB)! : u)
                onBase = true
            } else {
                onBase = false
            }
        }
        return String(out).split(separator: " ").joined(separator: "-")
    }

    // --------------------------------------------------------- localization

    /* koJosa lived here: the -(으)로서 allomorphy. Native-speaker review
    2026-08-25 confirmed the RULE was correct and the PARTICLE was the mistake.
    Removed with the particle it served. */

    // One template, every language: who, role, scope joined by spaces, scope
    // omitted when empty. Native-speaker review 2026-08-25 (47 reviews, 11 languages,
    // 7 models, 5 labs): bare apposition was proposed independently in every language,
    // and who-role-scope was unanimous in 8 of 11 and 35 of all 47. See coia.py for the
    // full rationale and for the connectors and allomorphy rules this replaced.
    static func template(_ w: String, _ r: String, _ s: String) -> String {
        return "\(w) \(r)" + (s.isEmpty ? "" : " \(s)")
    }

    static func pronoun(_ lang: String) -> String {
        ["en": "me", "es": "yo", "de": "ich", "fr": "moi", "pt": "eu", "it": "io",
         "ru": "я", "ja": "私", "zh": "我", "ko": "나", "ar": "أنا", "he": "אני"][lang] ?? ""
    }

    // ----------------------------------------------------------------- flags

    static let ASSIGNED: Set<Character> = ["0", "1", "4", "5", "6", "7", "8", "9"]

    struct CoiaError: Error { let message: String }

    static func flagGroup(_ digits: String, _ name: String) throws -> String {
        if digits.isEmpty { return "" }
        var seen = Set<Character>()
        for u in digits.unicodeScalars {
            guard let d = DIGITS[u.value] else {
                throw CoiaError(message: "\(name) must contain only decimal digits")
            }
            seen.insert(d)
        }
        return String(seen.sorted(by: >))
    }

    /// Mint an alias (§4, §6).
    static func createAlias(_ lang: String, _ who: String, _ role: String,
                            _ scope: String = "", _ flags: String = "",
                            _ privateFlags: String = "") throws -> String {
        let reflexive = (who == ME)
        let subject = reflexive ? pronoun(lang) : who
        let f = try flagGroup(flags, "flags")
        let pf = try flagGroup(privateFlags, "private_flags")
        for d in f where !ASSIGNED.contains(d) {
            throw CoiaError(message: "digit \(d) is reserved")
        }
        if reflexive && f.contains("0") {
            throw CoiaError(message: "reflexive aliases must not carry flag 0")
        }
        if pronoun(lang).isEmpty {
            throw CoiaError(message: "unsupported language \(lang)")
        }
        let expanded = template(subject, role, scope)
        let body = normalize(expanded)
        if normalize(role).isEmpty {
            throw CoiaError(message: "role must be non-empty after normalization")
        }
        if !reflexive && normalize(subject).isEmpty {
            throw CoiaError(message: "who must be non-empty after normalization")
        }
        if body.isEmpty {
            throw CoiaError(message: "alias is empty after normalization")
        }
        var out = body
        if !f.isEmpty || !pf.isEmpty { out += "," + f }
        if !pf.isEmpty { out += "," + pf }
        return out
    }

    /// Split a raw alias into body and flag groups (§6.1, §6.2).
    static func parseAlias(_ s: String) throws -> [String] {
        let unified = String(s.map { c -> Character in
            (c == "、" || c == "،" || c == "，") ? "," : c
        })
        let parts = unified.components(separatedBy: ",")
        if parts.count > 3 {
            throw CoiaError(message: "an alias has at most two flag groups")
        }
        return [normalize(parts[0]),
                try flagGroup(parts.count > 1 ? parts[1] : "", "flags"),
                try flagGroup(parts.count > 2 ? parts[2] : "", "private_flags")]
    }

    // -------------------------------------------------------------- matching

    static func terms(_ q: String) -> [String] {
        q.split(separator: "-").map(String.init)
    }

    /// Does a query find an alias? (§7)
    static func matches(_ query: String, _ alias: String) -> Bool {
        guard let body = try? parseAlias(alias)[0],
              let qbody = try? parseAlias(query)[0] else { return false }
        let ts = terms(qbody)
        return !ts.isEmpty && ts.allSatisfy { body.contains($0) }
    }

    /// Matching aliases in the normative order (§7).
    static func search(_ query: String, _ aliases: [String]) -> [String] {
        guard let qbody = try? parseAlias(query)[0] else { return [] }
        let ts = terms(qbody)
        var rows: [(Int, Int, Double, String)] = []
        for a in aliases {
            guard let body = try? parseAlias(a)[0] else { continue }
            if ts.isEmpty || !ts.allSatisfy({ body.contains($0) }) { continue }
            let segs = body.split(separator: "-").map(String.init)
            let whole = ts.filter { segs.contains($0) }.count
            let first = ts.compactMap { t -> Int? in
                guard let r = body.range(of: t) else { return nil }
                return body.distance(from: body.startIndex, to: r.lowerBound)
            }.min() ?? 0
            let matched = ts.reduce(0) { $0 + $1.unicodeScalars.count }
            let cover = Double(matched) / Double(body.unicodeScalars.count)
            rows.append((-whole, first, -cover, a))
        }
        rows.sort {
            if $0.0 != $1.0 { return $0.0 < $1.0 }
            if $0.1 != $1.1 { return $0.1 < $1.1 }
            if $0.2 != $1.2 { return $0.2 < $1.2 }
            return $0.3 < $1.3
        }
        return rows.map { $0.3 }
    }
}

File: impl/go/coia.go
// COIA v2 reference implementation (Go).
// Tables and vectors are generated; see ../../gencode.py.
package coia

import (
	"fmt"
	"sort"
	"strings"
	"unicode"

	"golang.org/x/text/unicode/norm"
)

const (
	tatweel = 0x0640
	zwj     = 0x200D
	zwnj    = 0x200C
)

// ME is the reflexive sentinel, distinct from the empty string (§4.1).
const ME = "\x00COIA-ME"

var fold = map[rune]rune{0x02BC: 0x02BB}

func inRanges(r rune, rs [][2]rune) bool {
	for _, p := range rs {
		if r >= p[0] && r <= p[1] {
			return true
		}
	}
	return false
}

// Default Case Folding from the normative table (Appendix B.6). Deliberately does
// NOT use strings.ToLower, which applies SIMPLE case mapping: it turns U+0130 into
// "i" where full folding gives "i" followed by U+0307.
func casefold(s string) string {
	var b strings.Builder
	for _, r := range s {
		if f, ok := Foldgap[r]; ok {
			b.WriteString(f)
		} else {
			b.WriteRune(r)
		}
	}
	return b.String()
}

func joinerOk(rs []rune, i int) bool {
	switch rs[i] {
	case zwj:
		return i > 0 && inRanges(rs[i-1], Virama)
	case zwnj:
		if i > 0 && inRanges(rs[i-1], Virama) {
			return true
		}
		return i > 0 && i < len(rs)-1 && inRanges(rs[i-1], Arabic) && inRanges(rs[i+1], Arabic)
	}
	return false
}

// Normalize reduces a string to COIA canonical form (§5).
func Normalize(s string) string {
	rs := []rune(casefold(norm.NFKC.String(s)))
	var b strings.Builder
	onBase := false
	for i, r := range rs {
		switch {
		case inRanges(r, Split):
			b.WriteRune(' ')
			onBase = false
		case r == tatweel:
			// decorative
		case unicode.Is(unicode.Cf, r):
			if joinerOk(rs, i) {
				b.WriteRune(r)
			}
		case unicode.Is(unicode.Me, r):
			// enclosing marks are decorative
		case unicode.Is(unicode.Mn, r) || unicode.Is(unicode.Mc, r):
			if onBase {
				b.WriteRune(r)
			}
		case unicode.IsLetter(r) || unicode.IsNumber(r):
			if f, ok := fold[r]; ok {
				r = f
			}
			b.WriteRune(r)
			onBase = true
		default:
			onBase = false
		}
	}
	return strings.Join(strings.Fields(b.String()), "-")
}

// ------------------------------------------------------------- localization

// koJosa lived here: the -(으)로서 allomorphy. Native-speaker review
// 2026-08-25 confirmed the RULE was correct and the PARTICLE was the mistake.
// Removed with the particle it served.

// One template, every language: who, role, scope joined by spaces, scope
// omitted when empty. Native-speaker review 2026-08-25 (47 reviews, 11 languages,
// 7 models, 5 labs): bare apposition was proposed independently in every language,
// and who-role-scope was unanimous in 8 of 11 and 35 of all 47. See coia.py for the
// full rationale and for the connectors and allomorphy rules this replaced.
func template(w, r, s string) string {
	return w + " " + r + opt(s, " "+s)
}

func opt(cond, val string) string {
	if cond == "" {
		return ""
	}
	return val
}

var pronouns = map[string]string{
	"en": "me", "es": "yo", "de": "ich", "fr": "moi", "pt": "eu", "it": "io",
	"ru": "я", "ja": "私", "zh": "我", "ko": "나", "ar": "أنا", "he": "אני",
}

// -------------------------------------------------------------------- flags

var assigned = map[rune]bool{'0': true, '1': true, '4': true, '5': true,
	'6': true, '7': true, '8': true, '9': true}

func flagGroup(digits, name string) (string, error) {
	if digits == "" {
		return "", nil
	}
	seen := map[rune]bool{}
	for _, r := range digits {
		d, ok := Digits[r]
		if !ok {
			return "", fmt.Errorf("%s must contain only decimal digits, got %q", name, r)
		}
		seen[d] = true
	}
	out := make([]rune, 0, len(seen))
	for d := range seen {
		out = append(out, d)
	}
	sort.Slice(out, func(i, j int) bool { return out[i] > out[j] })
	return string(out), nil
}

// CreateAlias mints an alias (§4, §6).
func CreateAlias(lang, who, role, scope, flags, privateFlags string) (string, error) {
	if _, ok := pronouns[lang]; !ok {
		return "", fmt.Errorf("unsupported language %q", lang)
	}
	reflexive := who == ME
	if reflexive {
		who = pronouns[lang]
	}
	f, err := flagGroup(flags, "flags")
	if err != nil {
		return "", err
	}
	pf, err := flagGroup(privateFlags, "private_flags")
	if err != nil {
		return "", err
	}
	for _, d := range f {
		if !assigned[d] {
			return "", fmt.Errorf("digit %q is reserved", d)
		}
	}
	if reflexive && strings.ContainsRune(f, '0') {
		return "", fmt.Errorf("reflexive aliases must not carry flag 0")
	}

	body := Normalize(template(who, role, scope))
	if Normalize(role) == "" {
		return "", fmt.Errorf("role must be non-empty after normalization")
	}
	if !reflexive && Normalize(who) == "" {
		return "", fmt.Errorf("who must be non-empty after normalization")
	}
	if body == "" {
		return "", fmt.Errorf("alias is empty after normalization")
	}

	out := body
	if f != "" || pf != "" {
		out += "," + f
	}
	if pf != "" {
		out += "," + pf
	}
	return out, nil
}

var commaVariants = strings.NewReplacer("、", ",", "،", ",", "，", ",")

// ParseAlias splits a raw alias into body and flag groups (§6.1, §6.2).
func ParseAlias(s string) (string, string, string, error) {
	parts := strings.Split(commaVariants.Replace(s), ",")
	if len(parts) > 3 {
		return "", "", "", fmt.Errorf("an alias has at most two flag groups")
	}
	get := func(i int) string {
		if i < len(parts) {
			return parts[i]
		}
		return ""
	}
	g1, err := flagGroup(get(1), "flags")
	if err != nil {
		return "", "", "", err
	}
	g2, err := flagGroup(get(2), "private_flags")
	if err != nil {
		return "", "", "", err
	}
	return Normalize(parts[0]), g1, g2, nil
}

// ----------------------------------------------------------------- matching

func terms(q string) []string {
	var out []string
	for _, t := range strings.Split(q, "-") {
		if t != "" {
			out = append(out, t)
		}
	}
	return out
}

// Matches reports whether a query finds an alias (§7).
func Matches(query, alias string) bool {
	body, _, _, err := ParseAlias(alias)
	if err != nil {
		return false
	}
	qbody, _, _, err := ParseAlias(query)
	if err != nil {
		return false
	}
	ts := terms(qbody)
	if len(ts) == 0 {
		return false
	}
	for _, t := range ts {
		if !strings.Contains(body, t) {
			return false
		}
	}
	return true
}

// Search returns matching aliases in the normative order (§7).
func Search(query string, aliases []string) []string {
	qbody, _, _, err := ParseAlias(query)
	if err != nil {
		return nil
	}
	ts := terms(qbody)
	type row struct {
		whole int
		first int
		cover float64
		alias string
	}
	var rows []row
	for _, a := range aliases {
		body, _, _, err := ParseAlias(a)
		if err != nil || len(ts) == 0 {
			continue
		}
		ok := true
		for _, t := range ts {
			if !strings.Contains(body, t) {
				ok = false
				break
			}
		}
		if !ok {
			continue
		}
		segs := strings.Split(body, "-")
		whole, first, matched := 0, 1<<30, 0
		for _, t := range ts {
			for _, s := range segs {
				if s == t {
					whole++
					break
				}
			}
			if i := strings.Index(body, t); i < first {
				first = i
			}
			matched += len([]rune(t))
		}
		rows = append(rows, row{whole, first, float64(matched) / float64(len([]rune(body))), a})
	}
	sort.Slice(rows, func(i, j int) bool {
		a, b := rows[i], rows[j]
		if a.whole != b.whole {
			return a.whole > b.whole
		}
		if a.first != b.first {
			return a.first < b.first
		}
		if a.cover != b.cover {
			return a.cover > b.cover
		}
		return a.alias < b.alias
	})
	out := make([]string, len(rows))
	for i, r := range rows {
		out[i] = r.alias
	}
	return out
}
